csharp/src/Apache.Arrow/ChunkedArray.cs

```diff
@@ -0,0 +1,87 @@
+﻿// Licensed to the Apache Software Foundation (ASF) under one or more
+// contributor license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright ownership.
+// The ASF licenses this file to You under the Apache License, Version 2.0
+// (the "License"); you may not use this file except in compliance with
+// the License.  You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System;
+using System.Collections.Generic;
+using Apache.Arrow;
+using Apache.Arrow.Types;
+
+namespace Apache.Arrow
+{
+    /// <summary>
+    /// A data structure to manage a list of primitive Array arrays logically as one large array
+    /// </summary>
+    internal class ChunkedArray
+    {
+        public readonly IList<Array> Arrays;
+        public readonly IArrowType DataType;
+        public readonly long Length;
+        public readonly long NullCount;
+
+        public ChunkedArray(IList<Array> arrays)
+        {
+            Arrays = arrays ?? throw new ArgumentNullException(nameof(arrays));
+            if (arrays.Count < 1)
+            {
+                throw new ArgumentException($"Count must be at least 1. Got {arrays.Count} instead");
+            }
+            DataType = arrays[0].Data.DataType;
+            foreach (Array array in arrays)
+            {
+                Length += array.Length;
+                NullCount += array.NullCount;
+            }
+        }
+
+        public ChunkedArray(Array array) : this(new[] { array }) { }
+
+        public ChunkedArray Slice(long offset, long length)
+        {
+            int curArrayIndex = GetArrayContainingRowIndex(ref offset);
+
+            IList<Array> newArrays = new List<Array>();
+            var numArrays = Arrays.Count;
+            while (curArrayIndex < numArrays && length > 0)
+            {
+                newArrays.Add(Arrays[curArrayIndex].Slice((int)offset,
+                              length > Arrays[curArrayIndex].Length ? Arrays[curArrayIndex].Length : (int)length));
+                length -= Arrays[curArrayIndex].Length - offset;
+                offset = 0;
+                curArrayIndex++;
+            }
+            return new ChunkedArray(newArrays);
+        }
+        public ChunkedArray Slice(long offset)
+        {
+            return Slice(offset, Length);
+        }
+        private int GetArrayContainingRowIndex(ref long rowIndex)
```

Comment:
I've stated this in some of the corefxlab PRs, but I don't like this pattern of passing in a `ref long` and having it implicitly being understood that after the call, `rowIndex` is now guaranteed to be able to be casted to an `int`.  I think this function should take 2 parameters:

`private int GetArrayContainingRowIndex(long rowIndex, out int resultArrayIndex)`

Reply:
There's only 1 caller to this method, so I just inlined it. I don't know why the C++ lib doesn't just replace this with `arrayIndex = offset/Arrays[0].Length`? I haven't seen a case where `Arrays[2]` is populated before `Arrays[1]` is full yet, so that is a potential optimization for later. We just need to review that `Append`(if the function exists) has the right behavior.